Derivation on scala 3 fails for case classes with higher-kinded type parameters

Using scala 3, this code triggers an error at compile-time: https://scastie.scala-lang.org/mrdziuban/pGKWKNNeSCKQWVnM5mUJIw/1

```scala
import ru.tinkoff.phobos.decoding.XmlDecoder
import ru.tinkoff.phobos.derivation.semiauto.deriveXmlDecoder

case class Test[F[_]](fi: F[Int])

object Test {
  given xd: XmlDecoder[Test[Option]] = deriveXmlDecoder("Test")
}
```

```
Exception occurred while executing macro expansion.
java.lang.AssertionError: assertion failed: TypeBounds(TypeRef(ThisType(TypeRef(NoPrefix,module class scala)),class Nothing),HKTypeLambda(List(_$1), List(TypeBounds(TypeRef(ThisType(TypeRef(NoPrefix,module class scala)),class Nothing),TypeRef(ThisType(TypeRef(NoPrefix,module class scala)),class Any))), TypeRef(ThisType(TypeRef(NoPrefix,module class scala)),class Any), List()))
  at scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
  at dotty.tools.dotc.core.Types$TypeBounds.<init>(Types.scala:5104)
  at dotty.tools.dotc.core.Types$RealTypeBounds.<init>(Types.scala:5180)
  at dotty.tools.dotc.core.Types$TypeBounds$.apply(Types.scala:5224)
  at dotty.tools.dotc.core.Types$TypeBounds.derivedTypeBounds(Types.scala:5112)
  at dotty.tools.dotc.core.ConstraintHandling.addOneBound(ConstraintHandling.scala:267)
  at dotty.tools.dotc.core.ConstraintHandling.addOneBound$(ConstraintHandling.scala:27)
  at dotty.tools.dotc.core.ProperGadtConstraint.addOneBound(GadtConstraint.scala:60)
  at dotty.tools.dotc.core.ConstraintHandling.addBoundTransitively(ConstraintHandling.scala:321)
  at dotty.tools.dotc.core.ConstraintHandling.addBoundTransitively$(ConstraintHandling.scala:27)
  at dotty.tools.dotc.core.ProperGadtConstraint.addBoundTransitively(GadtConstraint.scala:60)
  at dotty.tools.dotc.core.ProperGadtConstraint.addBound(GadtConstraint.scala:167)
  at dotty.tools.dotc.core.TypeComparer.gadtAddLowerBound(TypeComparer.scala:118)
  at dotty.tools.dotc.core.TypeComparer.narrowGADTBounds(TypeComparer.scala:1931)
  at dotty.tools.dotc.core.TypeComparer.compareGADT$1(TypeComparer.scala:523)
  at dotty.tools.dotc.core.TypeComparer.thirdTryNamed$1(TypeComparer.scala:526)
  at dotty.tools.dotc.core.TypeComparer.thirdTry$1(TypeComparer.scala:575)
  at dotty.tools.dotc.core.TypeComparer.secondTry$1(TypeComparer.scala:506)
  at dotty.tools.dotc.core.TypeComparer.compareNamed$1(TypeComparer.scala:306)
  at dotty.tools.dotc.core.TypeComparer.firstTry$1(TypeComparer.scala:312)
  at dotty.tools.dotc.core.TypeComparer.recur(TypeComparer.scala:1336)
  at dotty.tools.dotc.core.TypeComparer.isSubType(TypeComparer.scala:194)
  at dotty.tools.dotc.core.TypeComparer.isSubType(TypeComparer.scala:204)
  at dotty.tools.dotc.core.TypeComparer.topLevelSubType(TypeComparer.scala:128)
  at dotty.tools.dotc.core.TypeComparer$.topLevelSubType(TypeComparer.scala:2752)
  at dotty.tools.dotc.core.Types$Type.$less$colon$less(Types.scala:1044)
  at scala.quoted.runtime.impl.QuoteMatcher$.$eq$qmark$eq(QuoteMatcher.scala:345)
  at scala.quoted.runtime.impl.QuoteMatcher$.treeMatch(QuoteMatcher.scala:126)
  at scala.quoted.runtime.impl.QuotesImpl.scala$quoted$runtime$impl$QuotesImpl$$treeMatch(QuotesImpl.scala:3096)
  at scala.quoted.runtime.impl.QuotesImpl$TypeMatch$.unapply(QuotesImpl.scala:3066)
  at ru.tinkoff.phobos.derivation.decoder$.decodeElementCases$$anonfun$1(decoder.scala:113)
  at scala.collection.immutable.List.map(List.scala:246)
  at ru.tinkoff.phobos.derivation.decoder$.decodeElementCases(decoder.scala:130)
  at ru.tinkoff.phobos.derivation.decoder$.decodeStartElement(decoder.scala:141)
  at ru.tinkoff.phobos.derivation.decoder$.deriveProduct$$anonfun$1(decoder.scala:360)
  at ru.tinkoff.phobos.derivation.decoder$.deriveProduct$$anonfun$adapted$1(decoder.scala:370)
  at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:109)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1452)
  at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:134)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transformBlock(Trees.scala:1511)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1432)
  at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:134)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1434)
  at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:134)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transformBlock(Trees.scala:1511)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1432)
  at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:134)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1440)
  at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:134)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform$$anonfun$1(Trees.scala:1513)
  at scala.collection.immutable.List.mapConserve(List.scala:472)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1513)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transformSub(Trees.scala:1517)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1438)
  at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:134)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1434)
  at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:134)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transformBlock(Trees.scala:1511)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1432)
  at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:134)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1484)
  at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:134)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform$$anonfun$1(Trees.scala:1513)
  at scala.collection.immutable.List.mapConserve(List.scala:472)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1513)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transformStats(Trees.scala:1509)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transformBlock(Trees.scala:1511)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1432)
  at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:134)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1484)
  at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:134)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform$$anonfun$1(Trees.scala:1513)
  at scala.collection.immutable.List.mapConserve(List.scala:472)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1513)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transformStats(Trees.scala:1509)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1488)
  at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:134)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1486)
  at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:134)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform$$anonfun$1(Trees.scala:1513)
  at scala.collection.immutable.List.mapConserve(List.scala:472)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1513)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transformStats(Trees.scala:1509)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transformBlock(Trees.scala:1511)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1432)
  at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:134)
  at dotty.tools.dotc.quoted.PickledQuotes$.spliceTerms(PickledQuotes.scala:151)
  at dotty.tools.dotc.quoted.PickledQuotes$.unpickleTerm(PickledQuotes.scala:87)
  at scala.quoted.runtime.impl.QuotesImpl.unpickleExprV2(QuotesImpl.scala:3044)
  at ru.tinkoff.phobos.derivation.decoder$.deriveProduct(decoder.scala:370)
  at ru.tinkoff.phobos.derivation.decoder$.deriveElementDecoderImpl(decoder.scala:39)
  at ru.tinkoff.phobos.derivation.decoder$.deriveXmlDecoderImpl$$anonfun$1(decoder.scala:52)
  at ru.tinkoff.phobos.derivation.decoder$.deriveXmlDecoderImpl$$anonfun$adapted$1(decoder.scala:52)
  at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:109)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1452)
  at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:134)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform$$anonfun$1(Trees.scala:1513)
  at scala.collection.immutable.List.mapConserve(List.scala:472)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1513)
  at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1418)
  at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:134)
  at dotty.tools.dotc.quoted.PickledQuotes$.spliceTerms(PickledQuotes.scala:151)
  at dotty.tools.dotc.quoted.PickledQuotes$.unpickleTerm(PickledQuotes.scala:87)
  at scala.quoted.runtime.impl.QuotesImpl.unpickleExprV2(QuotesImpl.scala:3044)
  at ru.tinkoff.phobos.derivation.decoder$.deriveXmlDecoderImpl(decoder.scala:52)
```

--- a/README.md
+++ b/README.md
@@ -2,12 +2,12 @@
 [![Maven Central](https://img.shields.io/maven-central/v/ru.tinkoff/phobos-core_2.13.svg)](https://search.maven.org/search?q=ru.tinkoff.phobos-core)
 [![Build](https://github.com/TinkoffCreditSystems/phobos/actions/workflows/scala.yml/badge.svg)](https://github.com/TinkoffCreditSystems/phobos/actions/workflows/scala.yml)
 [![Scala Steward](https://img.shields.io/badge/Scala_Steward-helping-blue.svg?style=flat&logo=data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAA4AAAAQCAMAAAARSr4IAAAAVFBMVEUAAACHjojlOy5NWlrKzcYRKjGFjIbp293YycuLa3pYY2LSqql4f3pCUFTgSjNodYRmcXUsPD/NTTbjRS+2jomhgnzNc223cGvZS0HaSD0XLjbaSjElhIr+AAAAAXRSTlMAQObYZgAAAHlJREFUCNdNyosOwyAIhWHAQS1Vt7a77/3fcxxdmv0xwmckutAR1nkm4ggbyEcg/wWmlGLDAA3oL50xi6fk5ffZ3E2E3QfZDCcCN2YtbEWZt+Drc6u6rlqv7Uk0LdKqqr5rk2UCRXOk0vmQKGfc94nOJyQjouF9H/wCc9gECEYfONoAAAAASUVORK5CYII=)](https://scala-steward.org)
-[![Discord](https://img.shields.io/badge/chat-discord%20(en%2Fru)-blue)](https://discord.gg/S9Ad88t)
+[![Discord](https://img.shields.io/badge/chat-discord-blue)](https://discord.gg/S9Ad88t)
 
 Phobos is an XML data-binding library based on stream parsing. 
 It depends only on [aalto-xml](https://github.com/FasterXML/aalto-xml/) for parsing.
 
-Scala 2.12, 2.13 and 3.1 are supported. Scala 3.0 is supported in `core-3-0` module. See [Supported Scala versions](#supported-scala-versions) for more details.
+Scala 2.12, 2.13 and 3 are supported. See [Supported Scala versions](#supported-scala-versions) for more details.
 
 ## QuickStart
 Add phobos-core to your dependencies:
@@ -86,28 +86,24 @@ Where `<module>` is module name.
 
 
 ## Supported Scala versions
-Most modules support Scala 2.12, 2.13 and 3.1. 
+Most modules support Scala 2.12, 2.13 and 3. 
 Dependencies for some modules don't support Scala 3, thus these modules support only Scala 2.x.
-Most module dependencies are compiled for Scala 3.1, and they can not be used with Scala 3.0, 
-because Scala 3 [TASTy files are not forward compatible](https://contributors.scala-lang.org/t/improving-scala-3-forward-compatibility/5298).
-Module `core` is also compiled for Scala 3.1, but core features may be used via `core-3-0` module.
 
-Detailed information about supported Scala versions is in the table. Available versions for modules are marked with ✓.
+Detailed information about supported Scala versions is in the table below. Available versions for modules are marked with ✓.
 
-| Module name | 2.12 | 2.13 | 3.0 | 3.1 |
-|-------------|------|------|-----|-----|
-| core        |  ✓   |  ✓   |     |  ✓  |
-| core-3-0    |      |      |  ✓  |     |
-| akka-http   |  ✓   |  ✓   |     |     |
-| akka-stream |  ✓   |  ✓   |     |  ✓  |
-| ast         |  ✓   |  ✓   |     |     |
-| cats        |  ✓   |  ✓   |     |  ✓  |
-| derevo      |  ✓   |  ✓   |     |     |
-| enumeratum  |  ✓   |  ✓   |     |     |
-| fs2         |  ✓   |  ✓   |     |  ✓  |
-| fs2-ce2     |  ✓   |  ✓   |     |  ✓  |
-| monix       |  ✓   |  ✓   |     |  ✓  |
-| refined     |  ✓   |  ✓   |     |     |
+| Module name | 2.12 | 2.13 | 3 |
+|-------------|------|------|---|
+| core        |  ✓   |  ✓   | ✓ |
+| akka-http   |  ✓   |  ✓   |   |
+| akka-stream |  ✓   |  ✓   | ✓ |
+| ast         |  ✓   |  ✓   |   |
+| cats        |  ✓   |  ✓   | ✓ |
+| derevo      |  ✓   |  ✓   |   |
+| enumeratum  |  ✓   |  ✓   |   |
+| fs2         |  ✓   |  ✓   | ✓ |
+| fs2-ce2     |  ✓   |  ✓   | ✓ |
+| monix       |  ✓   |  ✓   | ✓ |
+| refined     |  ✓   |  ✓   |   |
 
 ## XSD and WSDL code-generation support
 Classes from XSD could be generated using [deimos](https://github.com/Tinkoff/deimos) library.

--- a/build.sbt
+++ b/build.sbt
@@ -53,19 +53,6 @@ lazy val `core` =
     .settings(commonDependencies)
     .jvmPlatform(scala3Versions)
 
-lazy val `core-3-0` =
-  (projectMatrix in file(s"modules/core"))
-    .settings(commonSettings("core-3-0"))
-    .settings(
-      target := file("modules/core/target30"),
-      libraryDependencies ++= List(
-        "com.fasterxml"  % "aalto-xml" % "1.3.2",
-        "org.scalatest" %% "scalatest" % "3.2.11" % "test",
-        "org.scalactic" %% "scalactic" % "3.2.11" % "test",
-      )
-    )
-    .jvmPlatform(List("3.0.2"))
-
 lazy val `akka-http` =
   (projectMatrix in file(s"modules/akka-http"))
     .settings(commonSettings("akka-http"))
@@ -196,7 +183,6 @@ lazy val `refined` =
 lazy val modules: Seq[ProjectReference] =
   Seq(
     `core`.projectRefs,
-    `core-3-0`.projectRefs,
     `akka-http`.projectRefs,
     `akka-stream`.projectRefs,
     `ast`.projectRefs,

--- a/modules/core/src/main/scala-3/ru/tinkoff/phobos/derivation/common.scala
+++ b/modules/core/src/main/scala-3/ru/tinkoff/phobos/derivation/common.scala
@@ -39,8 +39,10 @@ object common {
     val classTypeRepr = TypeRepr.of[T]
     val classSymbol = classTypeRepr.typeSymbol
 
-    val fields = classSymbol.primaryConstructor.paramSymss.flatten.map { fieldSymbol =>
-      val fieldAnnotations = fieldSymbol.annotations.map(_.asExpr)
+    // Extracting first non-type parameter list. Size of this parameter list must be equal to size of .caseFields
+    val constructorFields = classSymbol.primaryConstructor.paramSymss.filterNot(_.exists(_.isType)).head
+    val fields = classSymbol.caseFields.zip(constructorFields).map { (fieldSymbol, constructorFieldSymbol) =>
+      val fieldAnnotations = constructorFieldSymbol.annotations.map(_.asExpr)
       val fieldCategory    = extractFieldCategory(classSymbol, fieldSymbol, fieldAnnotations)
       val fieldXmlName     = extractFieldXmlName(config, classSymbol, fieldSymbol, fieldAnnotations, fieldCategory)
       val fieldNamespace   = extractFeildNamespace(config, classSymbol, fieldSymbol, fieldAnnotations, fieldCategory)

--- a/modules/core/src/main/scala-3/ru/tinkoff/phobos/derivation/decoder.scala
+++ b/modules/core/src/main/scala-3/ru/tinkoff/phobos/derivation/decoder.scala
@@ -188,10 +188,16 @@ object decoder {
     '{$c.getLocalName match {
       case name if name == $localName =>
         ${
+          val classTypeRepr = TypeRepr.of[T]
+          val primaryConstructor = Select(New(TypeTree.of[T]), classTypeRepr.typeSymbol.primaryConstructor)
+          val primaryConstructorTypeApplied = classTypeRepr match {
+            case AppliedType(_, params) => TypeApply(primaryConstructor, params.map(Inferred.apply))
+            case _ => primaryConstructor
+          }
           fields
-            .foldLeft[List[Term] => Expr[Either[DecodingError, T]]](
-              terms => '{Right(${Apply(Select(New(TypeTree.of[T]), TypeRepr.of[T].typeSymbol.primaryConstructor), terms).asExprOf[T]})}
-            ) { (acc, field) => params =>
+            .foldLeft[List[Term] => Expr[Either[DecodingError, T]]]{
+              terms => '{Right(${Apply(primaryConstructorTypeApplied, terms).asExprOf[T]})}
+            } { (acc, field) => params =>
               field.typeRepr.asType match {
                 case '[t] =>
                   val fSymbol = Symbol.newMethod(

--- a/modules/core/src/test/scala/ru/tinkoff/phobos/DecoderDerivationTest.scala
+++ b/modules/core/src/test/scala/ru/tinkoff/phobos/DecoderDerivationTest.scala
@@ -681,6 +681,84 @@ class DecoderDerivationTest extends AnyWordSpec with Matchers {
 
     "decode with default decoders sync" in decodeWithDefaultDecoders(pure)
     "decode with default decoders async" in decodeWithDefaultDecoders(fromIterable)
+
+    def workForHigherKindedData(toList: String => List[Array[Byte]]): Assertion = {
+      case class Foo[F[_]](a: F[Int], b: F[String], c: F[Int])
+      implicit val fooOptionDecoder: XmlDecoder[Foo[Option]] = deriveXmlDecoder("foo")
+      implicit val fooListDecoder: XmlDecoder[Foo[List]]     = deriveXmlDecoder("foo")
+
+      val fooOption = Foo[Option](Some(123), Some("b value"), None)
+      val fooList   = Foo[List](List(123), List("b value 1", "b value 2", "b value 3"), Nil)
+
+      val fooOptionString =
+        """<?xml version='1.0' encoding='UTF-8'?>
+          | <foo>
+          |   <a>123</a>
+          |   <b>b value</b>
+          | </foo>
+        """.stripMargin
+      val fooListString =
+        """<?xml version='1.0' encoding='UTF-8'?>
+          | <foo>
+          |   <a>123</a>
+          |   <b>b value 1</b>
+          |   <b>b value 2</b>
+          |   <b>b value 3</b>
+          | </foo>
+        """.stripMargin
+
+      assert(fooOptionDecoder.decodeFromIterable(toList(fooOptionString)) == Right(fooOption))
+      assert(fooListDecoder.decodeFromIterable(toList(fooListString)) == Right(fooList))
+    }
+
+    "work for higher-kinded data sync" in workForHigherKindedData(pure)
+    "work for higher-kinded data async" in workForHigherKindedData(fromIterable)
+
+    def workForNestedHigherKindedData(toList: String => List[Array[Byte]]): Assertion = {
+      case class Foo[F[_]](a: F[Int], b: F[String], c: F[Int])
+      implicit val fooOptionDecoder: ElementDecoder[Foo[Option]] = deriveElementDecoder
+      implicit val fooListDecoder: ElementDecoder[Foo[List]]     = deriveElementDecoder
+
+      case class Bar[F[_], G[_]](@attr qux: G[String], foo: F[Foo[F]])
+      implicit val barOptionDecoder: XmlDecoder[Bar[Option, Option]] = deriveXmlDecoder("bar")
+      implicit val barListDecoder: XmlDecoder[Bar[List, Option]]     = deriveXmlDecoder("bar")
+
+      val barOption =
+        Bar[Option, Option](Some("qux value"), Some(Foo[Option](Some(123), Some("b value"), None)))
+      val barList =
+        Bar[List, Option](
+          Some("qux value"),
+          List(Foo[List](List(123), List("b value 1", "b value 2", "b value 3"), Nil))
+        )
+
+      val barOptionString =
+        """<?xml version='1.0' encoding='UTF-8'?>
+          | <bar qux="qux value">
+          |   <foo>
+          |     <a>123</a>
+          |     <b>b value</b>
+          |   </foo>
+          | </bar>
+            """.stripMargin
+
+      val barListString =
+        """<?xml version='1.0' encoding='UTF-8'?>
+          | <bar qux="qux value">
+          |   <foo>
+          |     <a>123</a>
+          |     <b>b value 1</b>
+          |     <b>b value 2</b>
+          |     <b>b value 3</b>
+          |   </foo>
+          | </bar>
+            """.stripMargin
+
+      assert(barOptionDecoder.decodeFromIterable(toList(barOptionString)) == Right(barOption))
+      assert(barListDecoder.decodeFromIterable(toList(barListString)) == Right(barList))
+    }
+
+    "work for nested higher-kinded data sync" in workForNestedHigherKindedData(pure)
+    "work for nested higher-kinded data async" in workForNestedHigherKindedData(fromIterable)
   }
 
   "Decoder derivation for sealed traits" should {
@@ -1464,11 +1542,11 @@ class DecoderDerivationTest extends AnyWordSpec with Matchers {
           f: Double,
           qux: List[Qux]
       )
-      implicit val fooEncoder: ElementDecoder[Foo] = deriveElementDecoder
+      implicit val fooDecoder: ElementDecoder[Foo] = deriveElementDecoder
 
       final case class Bar(@xmlns(tcs) a: Int, b: String, c: Double, foo: Foo)
       val config                               = ElementCodecConfig.default.withScopeDefaultNamespace(tkf)
-      implicit val barEncoder: XmlDecoder[Bar] = deriveXmlDecoderConfigured("bar", config)
+      implicit val barDecoder: XmlDecoder[Bar] = deriveXmlDecoderConfigured("bar", config)
 
       val bar = Bar(123, "b value", 1.234, Foo(321, "e value", 4.321, List(Qux("g value 1", 1), Qux("g value 2", 2))))
       val string1 =
@@ -1538,11 +1616,11 @@ class DecoderDerivationTest extends AnyWordSpec with Matchers {
           f: Double,
       )
       val fooConfig                                = ElementCodecConfig.default.withScopeDefaultNamespace(xmp)
-      implicit val fooEncoder: ElementDecoder[Foo] = deriveElementDecoderConfigured(fooConfig)
+      implicit val fooDecoder: ElementDecoder[Foo] = deriveElementDecoderConfigured(fooConfig)
 
       final case class Bar(@xmlns(tcs) a: Int, b: String, c: Double, foo: Foo)
       val barConfig                            = ElementCodecConfig.default.withScopeDefaultNamespace(tkf)
-      implicit val barEncoder: XmlDecoder[Bar] = deriveXmlDecoderConfigured("bar", barConfig)
+      implicit val barDecoder: XmlDecoder[Bar] = deriveXmlDecoderConfigured("bar", barConfig)
 
       val bar = Bar(123, "b value", 1.234, Foo(321, "e value", 4.321))
       val string1 =

--- a/modules/core/src/test/scala/ru/tinkoff/phobos/EncoderDerivationTest.scala
+++ b/modules/core/src/test/scala/ru/tinkoff/phobos/EncoderDerivationTest.scala
@@ -391,6 +391,78 @@ class EncoderDerivationTest extends AnyWordSpec with Matchers {
           """.stripMargin.minimized,
       )
     }
+
+    "work for higher kinded data" in {
+      case class Foo[F[_]](a: F[Int], b: F[String], c: F[Int])
+      implicit val fooOptionEncoder: XmlEncoder[Foo[Option]] = deriveXmlEncoder("foo")
+      implicit val fooListEncoder: XmlEncoder[Foo[List]] = deriveXmlEncoder("foo")
+
+      val fooOption = Foo[Option](Some(123), Some("b value"), None)
+      val fooList = Foo[List](List(123), List("b value 1", "b value 2", "b value 3"), Nil)
+
+      val fooOptionString =
+        """<?xml version='1.0' encoding='UTF-8'?>
+          | <foo>
+          |   <a>123</a>
+          |   <b>b value</b>
+          | </foo>
+        """.stripMargin.minimized
+      val fooListString =
+        """<?xml version='1.0' encoding='UTF-8'?>
+          | <foo>
+          |   <a>123</a>
+          |   <b>b value 1</b>
+          |   <b>b value 2</b>
+          |   <b>b value 3</b>
+          | </foo>
+        """.stripMargin.minimized
+
+      assert(fooOptionEncoder.encode(fooOption) == fooOptionString)
+      assert(fooListEncoder.encode(fooList) == fooListString)
+    }
+
+    "work for nested higher-kinded data" in {
+      case class Foo[F[_]](a: F[Int], b: F[String], c: F[Int])
+      implicit val fooOptionEncoder: ElementEncoder[Foo[Option]] = deriveElementEncoder
+      implicit val fooListEncoder: ElementEncoder[Foo[List]] = deriveElementEncoder
+
+      case class Bar[F[_], G[_]](@attr qux: G[String], foo: F[Foo[F]])
+      implicit val barOptionEncoder: XmlEncoder[Bar[Option, Option]] = deriveXmlEncoder("bar")
+      implicit val barListEncoder: XmlEncoder[Bar[List, Option]] = deriveXmlEncoder("bar")
+
+      val barOption =
+        Bar[Option, Option](Some("qux value"), Some(Foo[Option](Some(123), Some("b value"), None)))
+      val barList =
+        Bar[List, Option](
+          Some("qux value"),
+          List(Foo[List](List(123), List("b value 1", "b value 2", "b value 3"), Nil))
+        )
+
+      val barOptionString =
+        """<?xml version='1.0' encoding='UTF-8'?>
+          | <bar qux="qux value">
+          |   <foo>
+          |     <a>123</a>
+          |     <b>b value</b>
+          |   </foo>
+          | </bar>
+        """.stripMargin.minimized
+
+      val barListString =
+        """<?xml version='1.0' encoding='UTF-8'?>
+          | <bar qux="qux value">
+          |   <foo>
+          |     <a>123</a>
+          |     <b>b value 1</b>
+          |     <b>b value 2</b>
+          |     <b>b value 3</b>
+          |   </foo>
+          | </bar>
+        """.stripMargin.minimized
+
+      assert(barOptionEncoder.encode(barOption) == barOptionString)
+      assert(barListEncoder.encode(barList) == barListString)
+    }
   }
 
   "Encoder derivation for sealed traits" should {

